bpf:trace: refactor L2/L3 packet check into classifiers

smagnani96 opened the pull request:
* Refactor current identical check in {trace,drop}.h for packets from L2/L3 devices into a new classifiers.h file.
* Moving also from using bits in `struct {trace,drop}_notify` to a new `__u8 flags` fields. This will help us in subsequent patches where we introduce additional classifiers.

Please refer to commit messages for further details.

kaworu reviewed (approved):
Awesome work @smagnani96 patch LGTM :tada: Before merging could you please open an issue to handle the ipsec/wg flag in Hubble so we can expose it in flows?

brb commented on bpf/lib/classifiers.h:
```diff
@@ -0,0 +1,43 @@
+	if (ctx->protocol == bpf_htons(ETH_P_IPV6))
+		return CLS_FLAG_L3_DEV | CLS_FLAG_IPV6;
+
+	return CLS_FLAG_L3_DEV;
```
What happens if a developer accidentally uses `ctx_device_classifiers()` on an L2 netdev? Have you considered adding `if (ETH_HLEN == 0) { return CLS_FLAG_L3_DEV; }`?

smagnani96 replied:
In theory, this returns `0` in case `IS_BPF_WIREGUARD` is not defined.
But yeah, could avoid the whole `#ifdef` logic and simply look at the `ETH_HLEN` 👍🏼

brb reviewed (commented):
I like the changes! However, one confusing bit is the API of the classifiers. At least to me it's not obvious when to use `ctx_device_classifiers` and when to use e.g., `ctx_from_netdev_...`?

smagnani96 committed:
bpf: refactor L2/L3 device check for tracing into classifiers

This commit refactors the current logic to signal packets coming from a
L3 device and IPv6 packets into a unique file `bpf/lib/classifiers.h` to
be used from within `bpf/lib/{trace,drop}.h`. Right now these two files
share the logic to detect such packets and set the apposite bit into their
message structures. With this commit we move from individual bits to a
unique `__u8 flags` field in the struct trace/drop. This helps with
upcoming patches where we'll add additional classifiers to also tag
Encrypted/Decrypted WireGuard/IPSec packets and overlay. This also aligns
to the userspace Hubble/Monitor, where we have a unique Flag field.

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

smagnani96 committed:
bpf: enable ctx_classify for bpf_host used in L3 devices

This commit extends the current ctx_classify logic introduced in the
previous commit to additionally account for the use case in which we
attach bpf_host to a L3 device, where ETH_HLEN is equal to zero.
We do not introduce though an additional bpf test, as this is already
covered by the WireGuard use case (classifiers_l2_dev.c).

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>
